File: elixir-elasticemail-examples/examples/webhooks.exs
alias ElasticEmail.Env

public_url = Env.public_url()
token = Env.webhook_token()

# Elastic Email does not sign webhook requests. The examples append a shared secret
# as a query parameter and the receiving handler checks it.
webhook_url = "#{public_url}/webhook?token=#{URI.encode_www_form(token)}"

fail = fn step, error ->
  IO.puts(:stderr, "Error (#{step}): #{ElasticEmail.format_error(error)}")
  System.halt(1)
end

# 1. Create. Elastic Email sends a GET to the URL on save, so the handler must be reachable
#    and answer 2xx (use a tunnel such as ngrok for local development).
webhook_id =
  case ElasticEmail.post("/webhook", %{
         "Name" => "examples-webhook",
         "URL" => webhook_url,
         "NotifyOncePerEmail" => false,
         "NotificationForSent" => true,
         "NotificationForOpened" => true,
         "NotificationForClicked" => true,
         "NotificationForUnsubscribed" => true,
         "NotificationForAbuseReport" => true,
         "NotificationForError" => true
       }) do
    {:ok, w} ->
      IO.puts("Webhook created: #{w["WebhookID"]} #{w["URL"]}")
      w["WebhookID"]

    error ->
      fail.("create webhook", error)
  end

# 2. List
case ElasticEmail.get("/webhook", limit: 50, offset: 0) do
  {:ok, webhooks} ->
    IO.puts("\nWebhooks (#{length(webhooks)}):")

    Enum.each(webhooks, fn w ->
      IO.puts(" - #{w["WebhookID"]} #{w["Name"]} #{w["URL"]} enabled=#{w["IsEnabled"]}")
    end)

  error ->
    fail.("list webhooks", error)
end

# 3. Delete the one we created (comment out to keep it)
if webhook_id do
  case ElasticEmail.delete("/webhook/#{ElasticEmail.path_segment(webhook_id)}") do
    {:ok, _} -> IO.puts("\nWebhook deleted: #{webhook_id}")
    error -> fail.("delete webhook", error)
  end
end

File: ruby-elasticemail-examples/examples/webhooks.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

##
# Webhook Management
#
# Elastic Email does not sign webhook requests. The examples append a shared secret
# as a query parameter and the receiving handler checks it.
#
# Usage: ruby examples/webhooks.rb

require_relative "ee"
require "erb"

webhook_api = ElasticEmail::WebhookApi.new

webhook_url = "#{PUBLIC_URL}/webhook?token=#{ERB::Util.url_encode(WEBHOOK_TOKEN)}"

# 1. Create. Elastic Email sends a GET to the URL on save, so the handler must be reachable
#    and answer 2xx (use a tunnel such as ngrok for local development).
webhook_id = nil
begin
  webhook = webhook_api.webhook_post(
    ElasticEmail::WebhookCreatePayload.new(
      name: "examples-webhook",
      url: webhook_url,
      notify_once_per_email: false,
      notification_for_sent: true,
      notification_for_opened: true,
      notification_for_clicked: true,
      notification_for_unsubscribed: true,
      notification_for_abuse_report: true,
      notification_for_error: true
    )
  )
  webhook_id = webhook.webhook_id
  puts "Webhook created: #{webhook_id} #{webhook.url}"
rescue ElasticEmail::ApiError => e
  api_error("create webhook", e)
end

# 2. List
begin
  webhooks = webhook_api.webhook_get(limit: 50, offset: 0)
  puts "\nWebhooks (#{webhooks.length}):"
  webhooks.each do |w|
    puts " - #{w.webhook_id} #{w.name} #{w.url} enabled=#{w.is_enabled}"
  end
rescue ElasticEmail::ApiError => e
  api_error("list webhooks", e)
end

# 3. Delete the one we created (comment out to keep it)
if webhook_id
  begin
    webhook_api.webhook_by_publicid_delete(webhook_id)
    puts "\nWebhook deleted: #{webhook_id}"
  rescue ElasticEmail::ApiError => e
    api_error("delete webhook", e)
  end
end
